app/frontend/src/components/designer/FormDesigner.vue

```diff
@@ -369,15 +443,135 @@ export default {
       if (!this.isDirty) this.setDirtyFlag(true);
+
+      // If the form changed but was not done so through the undo
+      // or redo button
+      if (!this.patchHistoryUndo && !this.patchHistoryRedo) {
+        // flags and modified are defined when a component is added
+        if (_flags !== undefined && _modified !== undefined) {
+          if (this.componentAdded !== null) {
+            this.addPatchToHistory();
+          } else {
+            this.resetHistoryFlags();
+          }
+        } else {
+          if (this.componentRemoved) {
+            // Component was removed..
+            this.addPatchToHistory();
+          } else if (this.componentMoved) {
+            this.addPatchToHistory();
+          } else {
+            this.resetHistoryFlags();
+          }
+        }
+      } else {
+        // We pressed undo or redo, so we just ignore
+        // adding the action to the history
+        this.patchHistoryUndo = false;
+        this.patchHistoryRedo = false;
+        this.resetHistoryFlags();
+      }
+
+      this.undoEnabled = this.canUndoPatch();
+      this.redoEnabled = this.canRedoPatch();
     },
     onRenderMethod() {
       const el = document.querySelector('input.builder-sidebar_search:focus');
       if (el && el.value === '') this.reRenderFormIo += 1;
+      this.setDirtyFlag(false);
+
+      this.undoEnabled = this.canUndoPatch();
+      this.redoEnabled = this.canRedoPatch();
+    },
+    onAddComponentMethod(_info, _parent, _path, _index, _isNew) {
+      if (_isNew) {
+        // Component Add Start, the user can still cancel/remove the add
+        this.componentAdded = true;
+      } else {
+        // The user has initiated a move
+        this.componentMoved = true;
+      }
+    },
+    onRemoveComponentMethod(_component, _schema, _path, _index) {
+      // Component remove start
+      this.componentRemoved = true;
+    },
+    onUpdateComponentMethod(_component) {
+      this.componentUpdated = true;
+    },
+    // ----------------------------------------------------------------------------------/ FormIO Handlers
+
+    // ---------------------------------------------------------------------------------------------------
+    // Patch History
+    // ---------------------------------------------------------------------------------------------------
+    addPatchToHistory() {
+      // Remove any actions past the action we were on
+      if (this.patchHistory.length > 0) {
+        this.patchHistory.length = this.patchIndex + 1;
+      }
+
+      // Get the differences between the last patch 
+      // and the current form
+      var form = this.getPatch(++this.patchIndex);
+      var patch = compare(form, this.formSchema);
+      // Add the patch to the history
+      this.patchHistory.push(patch);
+
+      this.resetHistoryFlags();
+    },
+    getPatch(_idx) {
+      // Generate the form from the original schema
+      var form = JSON.parse(JSON.stringify(this.$options.originalSchema));
```

Comment:
- Do not use var. Declare with either const or let only.
- If I understand this correctly, you can use deepClone to achieve the same thing instead of stringify/parse chaining it.